src/DynamoCore/Graph/Nodes/CodeBlockNode.cs

```diff
@@ -627,30 +643,122 @@ internal void ProcessCodeDirect()
+            {
+                libraryServices.LibraryManagementCore.SetFunctionInactive(funcDef);
+            }
+        }
+
+        /// <summary>
+        /// This method skips parsing and only recompiles a CBN in a second pass after all other function definitions
+        /// have been compiled. It re-uses the ParseParam (ASTs etc.) of the CBN compiled from the first pass.
+        /// </summary>
+        /// <param name="errorMessage"></param>
+        /// <param name="warningMessage"></param>
+        internal void RecompileCodeBlockAST(ref string errorMessage, ref string warningMessage)
+        {
+            BuildStatus buildStatus = null;
+            try
+            {
+                int blockId = Constants.kInvalidIndex;
+                var core = libraryServices.LibraryManagementCore;
+
+                bool parsingPreloadFlag = core.IsParsingPreloadedAssembly;
+                bool parsingCbnFlag = core.IsParsingPreloadedAssembly;
+
+                core.IsParsingPreloadedAssembly = false;
+                core.IsParsingCodeBlockNode = true;
+                // This is always the second pass of precompiling a CBN so we need to reset this flag to false here.
+                core.IsCodeBlockNodeFirstPass = false;
+
+                core.ResetForPrecompilation();
+
+                // FunctionDefinitonNodes have already been compiled.
+                // Recompile the rest of the ASTs with function definitions in place.
+                var astNodes = ParseParam.ParsedNodes.Where(x => !(x is FunctionDefinitionNode));
+
+                // Clone a disposable copy of AST nodes for PreCompile() as Codegen mutates AST's
+                // while performing SSA transforms and we want to keep the original AST's
+                var codeblock = new CodeBlockNode();
+                var nodes = astNodes.Select(NodeUtils.Clone).ToList();
+                codeblock.Body.AddRange(nodes);
+
+                buildStatus = CompilerUtils.PreCompile(string.Empty, core, codeblock, out blockId);
+
+                core.IsParsingCodeBlockNode = parsingCbnFlag;
+                core.IsParsingPreloadedAssembly = parsingPreloadFlag;
+
+                ParseParam.AppendErrors(buildStatus.Errors);
+                ParseParam.AppendWarnings(buildStatus.Warnings);
+            }
+            catch (Exception)
+            {
+                buildStatus = null;
+            }
+
+            if (ParseParam.Errors.Any())
+            {
+                errorMessage = string.Join("\n", ParseParam.Errors.Select(m => m.Message));
+                ProcessError();
+                return;
+            }
+            if (ParseParam.Warnings != null)
+            {
+                // Unbound identifiers in CBN will have input slots.
+                // 
+                // To check function redefinition, we need to check other
+                // CBN to find out if it has been defined yet. Now just
+                // skip this warning.
```

Comment:
should this be a todo/followup jira?

Reply:
https://jira.autodesk.com/browse/DYN-4104